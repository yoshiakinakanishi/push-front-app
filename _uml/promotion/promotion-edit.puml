@startuml
title promotion edit task

actor Client <<user>>
participant View
control Task
collections Reducer
database State

group Initialize form behavior
    Client -> View: GET promotion/edit&promotionId=XXXXXX
    View -> Task: next@getInitialProps\nGET_PROMOTION, query.promotionId
    hnote over Task #82cddd : hundleRequestGetPromotion
    activate Task #ef857d
    Task -> API: saga@call\nGET /v2/promotions/${query.promotionId}
    alt Success 200
        API -> Task: response status 200
        Task -> Task: response.json()
        Task -> Reducer: saga@put\n(SET_PROMOTION, json)
        deactivate Task
        Reducer -> State: state@promotion\npromotionItem
        hnote over State #fffacd: promotion.promotionItem
        State -> State: object@assign\npromotionItem
        State -> View: mount initialValues: state.promotion.promotionItem
        View -> Client: GET 200 promotion/edit&promotionId=XXXXXX
    else No Content 204
        API -> Task: response status 204
        note left
            んーそもそもこういう事態ってあるのか。。。
        endnote
    else WIP faild 400
        API -> Task: error 400
        note left
            検討中
        endnote
    else WIP faild 401
        API -> Task: error 401
    else WIP faild 500
        API -> Task: error 500
    end
end

group Input form behavior
    Client -> View: promotion/edit&promotionId=XXXXXX\nInput update data
    View -> Reducer: actions@redux-form
    Reducer -> State: reducer@redux-form
    hnote over State #fffacd: form.promotionEdit
    State -> View: reduxForm@redux-form
    View -> Client: Change form view
end

group Submit form behavior
    Client -> View: promotion/edit&promotionId=XXXXXX\nClick submit <button />
    View -> Task: onSubmit\nactions@putPromotionUpdateData(params)\nPUT_PROMOTION_DATA,params
    note left
        redux-form側の
        バリデーション(invalid)を
        有効にすることで
        バリデーションがなくなるまで
        submitできない
    endnote
    activate Task #ef857d
    hnote over Task #a3d6cc : hundleRequestPutPromotionUpdateData
    Task -> View: saga@put\nstartSubmit
    activate View #ffe9a9
    hnote over View #ead7a4 :『登録中』の表示
    Task -> API: saga@call\nPUT /v2/promotions/${params.promotion_id}\nbody: JSON.stringify(params)
    alt Success 200
        API -> Task: response status 200
        Task -> View: saga@put\nstopSubmit
        deactivate View
        Task -> View : saga@put\nreset
        hnote over View #ead7a4 : form要素のリセット
        Task -> Reducer: saga@put\nDELETE_PROMOTION
        Reducer -> State: state@promotion\npromotionItem
        hnote over State #fffacd: promotion.promotionItem
        State -> State: object@assign\npromotionList => emply object
        hnote over State #ead7a4 : 更新前のStateを削除\n（initialValuesにマウントさせたプロモーション登録データ）
        Task -> View: next-routes@pushRoutes\npromotion/list
        deactivate Task
        hnote over View #ead7a4 : プロモーション一覧へ遷移
        View -> Client: GET 200 promotion/list
    else WIP faild 400
        API -> Task: error 400
        hnote over Task #e8383d: saga@put\nstopSubmit, error
        note left
            検討中
        endnote
    else WIP faild 401
        API -> Task: error 401
    else WIP faild 500
        API -> Task: error 500
    end
end

group Submit Cancel form behavior
    Client -> View: promotion/regist\nClick cancel <button />
    View -> Task: onSubmit\nactions@cancelPPutPromotionUpdateData\nCANCEL_PUT_PROMOTION_DATA
    activate Task #ef857d
    Task -> Reducer : saga@put\nreset
    Reducer -> State : redux-form@reset
    State -> View : form要素のリセット
    Task -> Reducer: saga@put\nDELETE_PROMOTION
    Reducer -> State: state@promotion\npromotionItem
    hnote over State #fffacd: promotion.promotionItem
    State -> State: object@assign\npromotionItem => emply object
    hnote over State #ead7a4 : 更新前のStateを削除\n（initialValuesにマウントさせたプロモーション登録データ）
    Task -> View: next-routes@pushRoutes\npromotion/list
    deactivate Task
    hnote over View #ead7a4 : プロモーション一覧へ遷移
    View -> Client: GET 200 promotion/list
end

@enduml