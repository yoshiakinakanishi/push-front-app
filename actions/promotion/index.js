import {
    GET_PROMOTION,
    SET_PROMOTION,
    GET_PROMOTION_LIST,
    SET_PROMOTION_LIST,
    DELETE_PROMOTION,
    POST_PROMOTION_DATA,
    CANCEL_POST_PROMOTION_DATA,
    PUT_PROMOTION_DATA,
    CANCEL_PUT_PROMOTION_DATA
} from '../../constants';

export const getPromotion = () => {
    return {
        type: GET_PROMOTION
    };
};

export const setPromotion = promotion => {
    return {
        type: SET_PROMOTION,
        promotionItem: promotion
    };
};

export const deletePromotion = promotion => {
    return {
        type: DELETE_PROMOTION,
        promotionItem: promotion
    };
};

export const getPromotionList = () => {
    return {
        type: GET_PROMOTION_LIST
    };
};

export const setPromotionList = promotionList => {
    return {
        type: SET_PROMOTION_LIST,
        promotionList: promotionList
    };
};

export const postPromotionRegistData = params => {
    return {
        type: POST_PROMOTION_DATA,
        params: params
    };
};

export const cancelPostPromotionRegistData = () => {
    return {
        type: CANCEL_POST_PROMOTION_DATA
    };
};

export const putPromotionUpdateData = params => {
    return {
        type: PUT_PROMOTION_DATA,
        params: params
    };
};

export const cancelPutPromotionUpdateData = () => {
    return {
        type: CANCEL_PUT_PROMOTION_DATA
    };
};
