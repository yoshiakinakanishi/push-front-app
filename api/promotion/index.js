import fetch from 'isomorphic-unfetch';
import { API_FQDN, PROMOTIONS } from '../../constants';

export const getPromotion = promotionId => {
    const URL = `${API_FQDN}/${PROMOTIONS}/${promotionId}`;

    return fetch(URL, {
        mode: 'cors',
        headers: {
            'Content-Type': 'application/json'
        }
    });
};

export const getPromotionList = () => {
    const URL = `${API_FQDN}/${PROMOTIONS}`;

    return fetch(URL, {
        mode: 'cors',
        headers: {
            'Content-Type': 'application/json'
        }
    });
};

export const postPromotionData = params => {
    const URL = `${API_FQDN}/${PROMOTIONS}`;

    return fetch(URL, {
        mode: 'cors',
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(params)
    });
};

export const putPromotionData = (promotionId, params) => {
    const URL = `${API_FQDN}/${PROMOTIONS}/${promotionId}`;

    return fetch(URL, {
        mode: 'cors',
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(params)
    });
};

export const deletePromotion = promotionId => {
    const URL = `${API_FQDN}/${PROMOTIONS}/${promotionId}`;

    return fetch(URL, {
        mode: 'cors',
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json'
        }
    });
};
