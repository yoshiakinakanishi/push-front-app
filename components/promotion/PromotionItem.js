import * as React from 'react';
import css from 'styled-jsx/css';
import { compose, pure } from 'recompose';
import { Link } from '../../routes';

const promotionItemStyle = css`
    .promotion_item {
        position: relative;
        width: 60%;
        height: 280px;
        margin: 30px auto;
        background-color: #fff;
        border-radius: 2px;
        box-shadow: rgba(0, 0, 0, 0.12) 0px 1px 6px,
            rgba(0, 0, 0, 0.12) 0px 1px 4px;
    }
    .promotion_title {
        position: absolute;
        top: 20px;
        left: 180px;
        width: 500px;
        height: 80px;
        line-height: 80px;
        font-size: 30px;
        font-weight: bold;
        color: #404040;
    }
    .promotion_titleText {
        position: absolute;
        top: 90px;
        left: 180px;
        width: 500px;
        height: 80px;
        line-height: 80px;
        color: #808080;
    }
    .promotion_image {
        position: absolute;
        top: 15px;
        left: 40px;
        width: 100px;
        height: 100px;
        background-color: #fff;
        background-image: linear-gradient(
                45deg,
                #444 25%,
                transparent 25%,
                transparent 75%,
                #444 75%,
                #444
            ),
            linear-gradient(
                -45deg,
                #444 25%,
                transparent 25%,
                transparent 75%,
                #444 75%,
                #444
            );
        background-size: 25px 25px;
        border-radius: 50%;
    }
    .promotion_status {
        position: absolute;
        top: 135px;
        left: 40px;
        width: 100px;
        height: 10px;
        line-height: 10px;
        text-align: center;
        color: gray;
        font-size: 14px;
    }
    .promotion_statusText {
        position: absolute;
        top: 155px;
        left: 40px;
        width: 100px;
        height: 30px;
        line-height: 30px;
        border: 1px solid #00bcd4;
        border-radius: 2px;
        text-align: center;
        color: #00bcd4;
    }
    .promotion_innerLine {
        width: 100%;
        height: 200px;
        border-bottom: 1px solid #e0e0e0;
    }
    .promotion_scheduleBtn {
        position: absolute;
        top: 200px;
        left: 20px;
        width: 180px;
        height: 40px;
        line-height: 40px;
        margin: 20px;
        background-color: #ff528f;
        border-radius: 2px;
        text-align: center;
        color: #fff;
        font-size: 18px;
        box-shadow: rgba(0, 0, 0, 0.12) 0px 1px 6px,
            rgba(0, 0, 0, 0.12) 0px 1px 4px;
    }
    .promotion_scheduleBtn:hover {
        opacity: 0.7;
        transition: 0.3s;
        cursor: pointer;
    }
    .promotion_editBtn {
        position: absolute;
        top: 200px;
        right: 20px;
        width: 100px;
        height: 40px;
        line-height: 40px;
        margin: 20px;
        border-radius: 2px;
        text-align: center;
        color: #404040;
        font-size: 18px;
        box-shadow: rgba(0, 0, 0, 0.12) 0px 1px 6px,
            rgba(0, 0, 0, 0.12) 0px 1px 4px;
    }
    .promotion_editBtn:hover {
        background-color: lightgray;
        transition: 0.3s;
        cursor: pointer;
    }
`;

// promotionのAPIを渡している

const PromotionItem = ({ promotion }) => (
    <li>
        <div className="promotion_item">
            <h3 className="promotion_title" data-test="title">
                {promotion.name}
            </h3>
            <p className="promotion_titleText" data-test="titleText">
                {promotion.description}
            </p>
            <div className="promotion_image" data-test="image" />
            <div className="promotion_status" data-test="status">
                {promotion.status}
            </div>
            <div className="promotion_statusText" data-test="statusText">
                {promotion.status === 'active' && '有効'}
                {promotion.status === 'inactive' && '無効'}
            </div>
            <div className="promotion_innerLine" />
            <Link route={`/schedule?promotionId=${promotion.promotion_id}`}>
                <a className="promotion_scheduleBtn" data-test="scheduleBtn">
                    配信設定
                </a>
            </Link>
            <Link
                route={`/promotion/edit?promotionId=${promotion.promotion_id}`}
            >
                <a className="promotion_editBtn" data-test="editBtn">
                    編集
                </a>
            </Link>
        </div>

        <style jsx>{promotionItemStyle}</style>
    </li>
);

export default compose(pure)(PromotionItem);
