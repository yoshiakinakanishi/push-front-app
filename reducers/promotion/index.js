import { combineReducers } from 'redux';
import promotionItem from './promotionItem';
import promotionList from './promotionList';

export default combineReducers({
    promotionItem,
    promotionList
});

// 変数名を引数に入れて、変数をStateにしてくれる
// reducerのディレクトリ構成がStateの構成
// 静的ページはinitialStateしてるけど、SSRなのでページ遷移したときに起こるので、毎回初期化はしない
