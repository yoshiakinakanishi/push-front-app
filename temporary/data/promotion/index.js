export const promotionList = {
    data: [
        {
            promotion_id: 'TA0000015866e489df31f70002',
            name: '動画',
            description: '動画サービスキャンペーン情報',
            status: 'active',
            timezone: '+09:00',
            notification_keys: [
                {
                    browser_type: 'chrome',
                    key: 'a96df14f8b5ccb567eede45bdddf189f'
                }
            ],
            created_at: '2017-07-08T16:18:44+09:00',
            updated_at: '2017-07-08T16:18:44+09:00'
        },
        {
            promotion_id: 'TA0000015866e489df31f70003',
            name: '動画',
            description: '動画サービスキャンペーン情報',
            status: 'inactive',
            timezone: '+09:00',
            notification_keys: [
                {
                    browser_type: 'chrome',
                    key: 'a96df14f8b5ccb567eede45bdddf189f'
                }
            ],
            created_at: '2017-07-08T16:18:44+09:00',
            updated_at: '2017-07-08T16:18:44+09:00'
        }
    ]
};
