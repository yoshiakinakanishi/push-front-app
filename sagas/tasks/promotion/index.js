import { call, put, take, takeEvery } from 'redux-saga/effects';
import * as actions from '../../../actions/promotion';
import { startSubmit, stopSubmit, reset } from 'redux-form';

import * as API from '../../../api/promotion';
import { Router } from '../../../routes';

export function* runRequestGetPromotionList() {
    try {
        const response = yield call(API.getPromotionList);
        if (response.status === 200) {
            const json = yield response.json();
            yield put(actions.setPromotionList(json));
        } else if (response.status === 204) {
            yield put(actions.setPromotionList([]));
        }
    } catch (err) {
        yield console.log(err);
    }
}

export function* handleRequestGetPromotionList() {
    yield takeEvery('GET_PROMOTION_LIST', runRequestGetPromotionList);
}

export function* handleRequestPostPromotionRegistData() {
    while (true) {
        const { params } = yield take('POST_PROMOTION_DATA');
        yield put(startSubmit('promotionRegist'));

        try {
            const response = yield call(API.postPromotionData, params);
            if (response.status === 201) {
                yield put(stopSubmit('promotionRegist'));
                yield put(reset('promotionRegist'));
                Router.pushRoute('/promotion/list');
            }
        } catch (err) {
            yield console.log(err);
            yield put(stopSubmit('promotionRegist'));
            yield put(reset('promotionRegist'));
        }
    }
}

export function* handleCancelPostPromotionRegistData() {
    while (true) {
        yield take('CANCEL_POST_PROMOTION_DATA');
        yield put(reset('promotionRegist'));
        Router.pushRoute('/promotion/list');
    }
}

export function* handleRequestGetPromotion() {
    while (true) {
        const { promotionId } = yield take('GET_PROMOTION');
        try {
            const response = yield call(API.getPromotion, promotionId);
            if (response.status === 200) {
                const json = yield response.json();
                yield put(actions.setPromotion(json));
            } else if (response.status === 204) {
                /** WIP */
            }
        } catch (err) {
            yield console.log(err);
        }
    }
}

export function* handleRequestPutPromotionUpdateData() {
    while (true) {
        const { params } = yield take('PUT_PROMOTION_DATA');
        yield put(startSubmit('promotionEdit'));

        try {
            const response = yield call(
                API.putPromotionData,
                params.promotion_id,
                params
            );
            if (response.status === 200) {
                yield put(stopSubmit('promotionEdit'));
                yield put(reset('promotionEdit'));
                yield put(actions.deletePromotion({}));
                Router.pushRoute('/promotion/list');
            }
        } catch (err) {
            yield console.log(err);
            yield put(stopSubmit('promotionEdit'));
            yield put(reset('promotionEdit'));
        }
    }
}

export function* handleCancelPutPromotionUpdateData() {
    while (true) {
        yield take('CANCEL_PUT_PROMOTION_DATA');
        yield put(reset('promotionEdit'));
        yield put(actions.deletePromotion({}));
        Router.pushRoute('/promotion/list');
    }
}
