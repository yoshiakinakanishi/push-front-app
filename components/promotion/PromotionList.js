import * as React from 'react';
// import css from 'styled-jsx/css';
import { compose, pure } from 'recompose';
import PromotionItem from './PromotionItem';

const PromotionItemZero = () => (
    <li>まだプロモーションが登録されていません。登録をしてください。</li>
);
const PromotionList = ({ promotionList }) => (
    <ul className="promotion_list">
        {promotionList.data.length > 0 &&
            promotionList.data.map(promotion => (
                <PromotionItem
                    key={promotion.promotion_id}
                    promotion={promotion}
                />
            ))}
        {!promotionList.data.length && <PromotionItemZero />}
    </ul>
);

export default compose(pure)(PromotionList);
