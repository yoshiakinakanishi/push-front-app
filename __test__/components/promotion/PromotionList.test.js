/* eslint-env jest */

import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import PromotionList from '../../../components/promotion/PromotionList';
import PromotionItem from '../../../components/promotion/PromotionItem';
import { promotionList } from '../../../temporary/data/promotion';

describe('Component Testing', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <PromotionList promotionList={promotionList} />
        ).dive();
    });
    it('+++ render the DUMB component', () => {
        expect(wrapper.length).toEqual(1);
    });
    it('+++ should render two <PromotionItem /> components', () => {
        expect(wrapper.find(PromotionItem).length).toEqual(2);
    });
});

describe('Snapshot Testing', () => {
    it('PromotionList show test', () => {
        const component = renderer.create(
            <PromotionList promotionList={promotionList} />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
