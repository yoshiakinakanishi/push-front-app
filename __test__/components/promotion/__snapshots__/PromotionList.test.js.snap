// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing PromotionList show test 1`] = `
<ul
  className="promotion_list"
>
  <li
    className="jsx-1263913822 promotion_item"
  >
    <div
      className="jsx-1263913822 promotion_title"
      data-test="title"
    >
      動画
    </div>
    <div
      className="jsx-1263913822 promotion_titleText"
      data-test="titleText"
    >
      動画サービスキャンペーン情報
    </div>
    <div
      className="jsx-1263913822 promotion_image"
      data-test="image"
    />
    <div
      className="jsx-1263913822 promotion_status"
      data-test="status"
    >
      active
    </div>
    <div
      className="jsx-1263913822 promotion_statusText"
      data-test="statusText"
    >
      有効
    </div>
    <div
      className="jsx-1263913822 promotion_innerLine"
    />
    <a
      className="jsx-1263913822 promotion_scheduleBtn"
      data-test="scheduleBtn"
      href="/schedule?promotionID=TA0000015866e489df31f70002"
      onClick={[Function]}
    >
      配信設定
    </a>
    <a
      className="jsx-1263913822 promotion_editBtn"
      data-test="editBtn"
      href="/promotion/edit?promotionId=TA0000015866e489df31f70002"
      onClick={[Function]}
    >
      編集
    </a>
  </li>
  <li
    className="jsx-1263913822 promotion_item"
  >
    <div
      className="jsx-1263913822 promotion_title"
      data-test="title"
    >
      動画
    </div>
    <div
      className="jsx-1263913822 promotion_titleText"
      data-test="titleText"
    >
      動画サービスキャンペーン情報
    </div>
    <div
      className="jsx-1263913822 promotion_image"
      data-test="image"
    />
    <div
      className="jsx-1263913822 promotion_status"
      data-test="status"
    >
      inactive
    </div>
    <div
      className="jsx-1263913822 promotion_statusText"
      data-test="statusText"
    >
      無効
    </div>
    <div
      className="jsx-1263913822 promotion_innerLine"
    />
    <a
      className="jsx-1263913822 promotion_scheduleBtn"
      data-test="scheduleBtn"
      href="/schedule?promotionID=TA0000015866e489df31f70003"
      onClick={[Function]}
    >
      配信設定
    </a>
    <a
      className="jsx-1263913822 promotion_editBtn"
      data-test="editBtn"
      href="/promotion/edit?promotionId=TA0000015866e489df31f70003"
      onClick={[Function]}
    >
      編集
    </a>
  </li>
</ul>
`;
